src/cuminc.jl

```diff
@@ -0,0 +1,124 @@
+an unbiased estimate of the failure probability.
+
+The estimator is given by
+``
+\\hat{I}(t) = \\sum_{i: t_i < t} \\frac{d_i}{n_i} \\hat{S}_1(t) \\hat{S}_2(t)
+``
+where ``\\hat{S}_1(t)`` is the Kaplan-Meier estimator using the event of interest
+and ``\\hat{S}_2(t)`` is the K-M estimator using the competing event.
+
+### References
+
+* Kalbfleisch, J. D., and Prentice, R. L. (1980). *The Statistical Analysis of
+  Failure Time Data*. New York, NY: John Wiley.
+"""
+struct CumulativeIncidence{T<:Real} <: NonparametricEstimator
+    times::Vector{T}
+    nevents::Vector{Int}
+    ncensor::Vector{Int}
+    ncompete::Vector{Int}
+    natrisk::Vector{Int}
+    cuminc::Vector{Float64}
+end
+
+function _cuminc(tte::AbstractVector{T}, status::BitVector, compete::BitVector) where {T}
+    nobs = length(tte)
+    dᵢ = 0              # Number of observed events of interest at time t
+    cᵢ = 0              # Number of censored events of interest at time t
+    cr = 0              # Number censored for the competing event
+    rᵢ = 0              # Number of observed competing events at time t
+    nᵢ = nobs           # Number remaining at risk just before time t
+    nr = nobs           # Number remaining at risk for the competing event
+    km1 = 1.0           # Ŝ(t) for the event of interest
+    km2 = 1.0           # Ŝ(t) for the competing event
+    inc = 0.0           # Î(t)
+
+    times = T[]         # The set of unique event times
+    nevents = Int[]     # Total observed events of interest at each time
+    ncensor = Int[]     # Total censored events of interest at each time
+    ncompete = Int[]    # Total observed competing events at each time
+    natrisk = Int[]     # Number at risk at each time
+    cuminc = Float64[]  # Cumulative incidence estimates
+
+    t_prev = zero(T)
+
+    @inbounds for i = 1:nobs
+        t = tte[i]
+        s = status[i]
+        r = compete[i]
+        # Aggregate over tied times
+        if t == t_prev
+            dᵢ += s
+            cᵢ += !s
+            rᵢ += r
+            cr += !r
+            continue
+        elseif !iszero(t_prev)
+            km1 *= 1 - dᵢ / nᵢ
+            km2 *= 1 - rᵢ / nr
+            inc += km1 * km2 * dᵢ / nᵢ
+            push!(times, t_prev)
+            push!(nevents, dᵢ)
+            push!(ncensor, cᵢ)
+            push!(ncompete, rᵢ)
+            push!(natrisk, nᵢ)
+            push!(cuminc, inc)
+        end
+        nᵢ -= dᵢ + cᵢ
+        nr -= rᵢ + cr
+        dᵢ = s
+        cᵢ = !s
+        rᵢ = r
+        cr = !r
+        t_prev = t
+    end
+
+    push!(times, t_prev)
+    push!(nevents, dᵢ)
+    push!(ncensor, cᵢ)
+    push!(ncompete, rᵢ)
+    push!(natrisk, nᵢ)
+    push!(cuminc, inc)
+
+    return CumulativeIncidence{T}(times, nevents, ncensor, ncompete, natrisk, cuminc)
+end
+
+function StatsBase.fit(::Type{CumulativeIncidence},
+                       times::AbstractVector{T},
```

Comment:
Any plans for a generalized `CompetingEventTime` which would hold both the time and type of (non-)event?

Reply:
Hm, hadn't thought of that, that's an interesting idea. We may want to consider adding that.